system/lib/pthread/proxying.c

```diff
@@ -0,0 +1,329 @@
+
+// Not thread safe.
+static int task_queue_empty(task_queue* tasks) {
+  return tasks->head == tasks->tail;
+}
+
+// Not thread safe.
+static int task_queue_full(task_queue* tasks) {
+  return tasks->head == (tasks->tail + 1) % tasks->capacity;
+}
+
+// // Not thread safe. Returns 1 on success and 0 on failure.
+static int task_queue_grow(task_queue* tasks) {
+  // Allocate a larger task queue.
+  int new_capacity = tasks->capacity * 2;
+  task* new_tasks = malloc(sizeof(task) * new_capacity);
+  if (new_tasks == NULL) {
+    return 0;
+  }
+  // Copy the tasks such that the head of the queue is at the beginning of the
+  // buffer. There are two cases to handle: either the queue wraps around the
+  // end of the old buffer or it does not.
+  int queued_tasks;
+  if (tasks->head <= tasks->tail) {
+    // No wrap. Copy the tasks in one chunk.
+    queued_tasks = tasks->tail - tasks->head;
+    memcpy(new_tasks, &tasks->tasks[tasks->head], sizeof(task) * queued_tasks);
+  } else {
+    // Wrap. Copy `first_queued` tasks up to the end of the old buffer and
+    // `last_queued` tasks at the beginning of the old buffer.
+    int first_queued = tasks->capacity - tasks->head;
+    int last_queued = tasks->tail;
+    queued_tasks = first_queued + last_queued;
+    memcpy(new_tasks, &tasks->tasks[tasks->head], sizeof(task) * first_queued);
+    memcpy(new_tasks + first_queued, tasks->tasks, sizeof(task) * last_queued);
+  }
+  free(tasks->tasks);
+  tasks->tasks = new_tasks;
+  tasks->capacity = new_capacity;
+  tasks->head = 0;
+  tasks->tail = queued_tasks;
+  return 1;
+}
+
+// Not thread safe. Returns 1 on success and 0 on failure.
+static int task_queue_enqueue(task_queue* tasks, task t) {
+  if (task_queue_full(tasks) && !task_queue_grow(tasks)) {
+    return 0;
+  }
+  tasks->tasks[tasks->tail] = t;
+  tasks->tail = (tasks->tail + 1) % tasks->capacity;
+  return 1;
+}
+
+// Not thread safe.
+static task task_queue_dequeue(task_queue* tasks) {
+  task t = tasks->tasks[tasks->head];
+  tasks->head = (tasks->head + 1) % tasks->capacity;
+  return t;
+}
+
+struct em_proxying_queue {
+  // Protects all accesses to all task_queues.
+  pthread_mutex_t mutex;
+  // `size` task queues stored in an array of size `capacity`.
+  task_queue* task_queues;
+  int size;
+  int capacity;
+};
+
+static em_proxying_queue system_proxying_queue = {.mutex =
+                                                    PTHREAD_MUTEX_INITIALIZER,
+                                                  .task_queues = NULL,
+                                                  .size = 0,
+                                                  .capacity = 0};
+
+em_proxying_queue* emscripten_proxy_get_system_queue(void) {
+  return &system_proxying_queue;
+}
+
+em_proxying_queue* em_proxying_queue_create(void) {
+  em_proxying_queue* q = malloc(sizeof(em_proxying_queue));
+  if (q == NULL) {
+    return NULL;
+  }
+  *q = (em_proxying_queue){.mutex = PTHREAD_MUTEX_INITIALIZER,
```

Comment:
cast needed?

Reply:
Yes, this is another compound literal expression.

Reply:
Interesting..  I don't remember using the casts when I've used that syntax in the past... seems like the compiler should be able know.   

Like for example in this case there is no cast needed:


```
typedef struct {                                                                                        
  int a;                                                                         
  int b;                                                                         
} Foo;                                                                           
                                                                                 
Foo f = {.a = 1, .b = 2};  
```

Reply:
Yeah, that should be simple type inference since C doesn't have any sort of struct subtyping. I'm not sure what the rationale for requiring the type to be repeated is.